### PR TITLE
Indexer: Deprecate `storyIndexers` API, add foundation for `indexers`

### PR DESCRIPTION
**Pull request by JReinhold**
Works on #23457 

## What I did

- Deprecated existing `storyIndexers` in favor of `indexers` as the new entry for the new API
- Doesn't actually implement support for `indexers` yet, they are ignored. **Which is also why this doesn't target `next` - we don't want to deprecate the API before the replacement is ready.**
- Move tests for `StoryIndexGenerator` to `StoryIndexGenerator.deprecate.test.ts`. The plan is to copy the file to `StoryIndexGenerator.test.ts` that will have the same tests but with the new API instead (plus a bunch more)

Follow-up PR here will implement support for the new `IndexEntry` in `StoryIndexGenerator`, which should be enough to make this releasable. Actually changing our internal indexers can be done afterwards.

## Checklist

- [ ] Make sure your changes are tested (stories and/or unit, integration, or end-to-end tests)
- [ ] Make sure to add/update documentation regarding your changes
- [ ] If you are deprecating/removing a feature, make sure to update
      [MIGRATION.MD](https://github.com/storybookjs/storybook/blob/next/MIGRATION.md)

#### Maintainers

- [ ] When this PR is ready for testing, make sure to add `ci:normal`, `ci:merged` or `ci:daily` GH label to it to run a specific set of sandboxes. The particular set of sandboxes can be found in `code/lib/cli/src/sandbox-templates.ts`
- [ ] Make sure this PR contains **one** of the labels below.

`["cleanup", "BREAKING CHANGE", "feature request", "bug", "build", "documentation", "maintenance", "dependencies", "other"]`

### CODE REVIEW

File: code/lib/core-server/src/utils/StoryIndexGenerator.deprecated.test.ts [hunk 1]
```diff
@@ -44,19 +45,20 @@ const storiesMdxIndexer = async (fileName: string, opts: any) => {
...
-  ] as StoryIndexer[],
+  ],
+  indexers: [],
   storiesV2Compatibility: false,
   storyStoreV7: true,
   docs: { defaultName: 'docs', autodocs: false },
 };
 
-describe('StoryIndexGenerator', () => {
+describe('StoryIndexGenerator with deprecated indexer API', () => {
```
Comment (JReinhold) [inline code/lib/core-server/src/utils/StoryIndexGenerator.deprecated.test.ts:61]: also notice the file rename here.

File: code/lib/csf-tools/src/CsfFile.ts [hunk 2]
```diff
@@ -234,7 +234,8 @@ export class CsfFile {
     return node;
   }
 
-  parse() {
+  // TODO: is there a better way to type this correctly?
+  parse(): CsfFile & { meta: StaticMeta; stories: IndexedStory[] } {
```
Comment (JReinhold) [inline code/lib/csf-tools/src/CsfFile.ts:238]: Would like some input here @kasperpeulen @tmeasday

There's no new implementation, [I just added types to our existing indexers](https://github.com/storybookjs/storybook/pull/23582/files#diff-e3d50a086df937d7c8f41d5819f0e6d8368961c112866cc291cfe057a4fafe23R132-R133) which surfaced this type error.
The problem is, that:

1. Most indexers return the result `CsfFile.parse()` as the result of the indexer.
2. An indexer needs to return an `IndexedCsfFile`
3. `CsfFile.parse()` returns `CsfFile`, not `IndexedCsfFile`
4. The difference between the two is that:
	1. `meta` can be `undefined` in `CsfFile`, it can't in `IndexedCsfFile`
	2. `stories.*.name` can be `undefined` in `CsfFile` (`StaticStory`), it can't in `IndexedCsfFile` (`IndexedStory`)
5. The implementation of `CsfFile.parse()` ensures that both of these constraints are met, but TypeScript of course doesn't know that because 3.

My solution here makes TypeScript happy, but it doesn't feel like a type-safe way to solve this?

File: code/lib/types/src/modules/indexer.ts [hunk 3]
```diff
@@ -63,7 +98,35 @@ export type DocsIndexEntry = BaseIndexEntry & {
   type: 'docs';
 };
 
+/**
+ * @deprecated This type is deprecated and will be replaced with {@link NewIndexEntry} in 8.0.
+ */
 export type IndexEntry = StoryIndexEntry | DocsIndexEntry;
+export interface NewBaseIndexEntry {
```
Comment (JReinhold) [inline code/lib/types/src/modules/indexer.ts:105]: It's annoying that we have to temporarily introduce this new type, but it's the only non-breaking way I can think of. We'll replace `IndexEntry` with `NewIndexEntry` in 8.0.

If users migrate now, they'll migrate to `NewIndexEntry`, only to migrate back to `IndexEntry` in 8.0, which must be annoying.

Comment (tmeasday) [reply code/lib/types/src/modules/indexer.ts:105]: I think this type isn't an `IndexEntry`. And index entry is what ends up in `index.json`, so should for instance always include an `id`.

I think this type is like an `IndexEntryInput` or something?

PS -- tell me more about `key`. For CSF files we don't need that because it's expected that the loader will produce a `CSFFile` which is keyed on `storyId`. Are we expecting something different in this case?

File: code/lib/csf-tools/src/CsfFile.ts (hunk 2 above)
Comment (tmeasday) [reply code/lib/csf-tools/src/CsfFile.ts:238]: Is there a reason you didn't just change it to return an `IndexedCSFFile` (I assume that's what you meant?)

Comment (JReinhold) [reply code/lib/csf-tools/src/CsfFile.ts:238]: `IndexedCSFFile` has less properties than `CsfFile`, so other callsites that uses `parse()` won't accept that. But I can change it to `CsfFile & IndexedCSFFile` which should be okay I guess.

File: code/lib/types/src/modules/indexer.ts (hunk 3 above)
Comment (JReinhold) [reply code/lib/types/src/modules/indexer.ts:105]: > I think this type isn't an `IndexEntry`. And index entry is what ends up in `index.json`, so should for instance always include an `id`.
> 
> I think this type is like an `IndexEntryInput` or something?

I see what you mean, of course! I've changed it to that, then `IndexEntry` also won't be deprecated.

> PS -- tell me more about `key`. For CSF files we don't need that because it's expected that the loader will produce a `CSFFile` which is keyed on `storyId`. That is the loader responsible for mapping the pure ESM module exports to an object keyed on `storyId`, so we can look up the story annotations only knowing the `storyId`.
> 
> Are we expecting something different in this case?

This type is taken directly from the [Indexer RFC](https://github.com/storybookjs/storybook/discussions/23176#Implementation), which also shows example indexers that makes use of the `key` property.
I'm assuming that this is added to make indexers and story definitions more flexible, in that the export doesn't need to match the story ID, but I'd like @shilman to elaborate here.

Are you saying that if we introduce this, we need to make the annotation lookup smarter, because then it can't just rely on the `id`, it also needs to know the `key`, which I guess would mean we need the `key` in the index as well?

File: code/lib/csf-tools/src/CsfFile.ts (hunk 2 above)
Comment (tmeasday) [reply code/lib/csf-tools/src/CsfFile.ts]: I don't really know the difference between the two types but is another option to just add the fields to `CsfFile` if that's what's returned by the `CsfFile` library?

Comment (JReinhold) [reply code/lib/csf-tools/src/CsfFile.ts]: The issue is that `CsfFile` already have `meta` and `stories`, but they can be `undefined` because they are not set during construction, only after `parse()` is called.

File: code/lib/types/src/modules/indexer.ts (hunk 3 above)
Comment (tmeasday) [reply code/lib/types/src/modules/indexer.ts]: >  I guess would mean we need the key in the index as well?

Yes, if the plan is to read the export named `key` in the browser when constructing the `NormalizedCsfFile`. However, I'm not sure if that makes sense because I'm pretty sure you are going to need to transform the raw module exports to get something that looks like CSF anyway (either in a loader or a runtime transformer ala [this comment](https://github.com/storybookjs/storybook/discussions/23176#discussioncomment-6267478)).

If you are already transforming the file into CSF, why wouldn't you transform the export names to mean this wasn't required (ie. so they match up with the story ids)? Perhaps I am missing a use case though.

Comment (shilman) [reply code/lib/types/src/modules/indexer.ts]: It's been awhile since I wrote the RFC but I think the rationale was twofold.

First, the consumer of the data needs to have the export name to be able to reference the story:

```
export default { title: 'Button' };
export const Primary = {};
```

This would generate a story with ID `button--primary`. But we don't know if the export key is `Primary`, `primary`, or something else. I'm not sure how this is handled today, but I felt that just providing the key directly would be more straightforward.

Second, it is more convenient to provide the key and have the calling code generate the story ID automatically from the export. And if you provide the ID that can take precedence.

If it's causing problems, I'm fine with using something closer to the current interface instead. But I'd like to know how it actually works currently, since it seems like it would be a bit convoluted.

Comment (JReinhold) [reply code/lib/types/src/modules/indexer.ts]: I remember @tobiasdiez having [issues with generating exports in the Vue CSF addon](https://github.com/tobiasdiez/storybook-vue-addon/pull/41), because he wanted a story named `default` (pretty sensible), but that was a reserved keyword, so he had to uppercase everything to get `Default`. In that scenario, if he was able to specify the key independent of the ID, he could prepend all exports and make them like `export const vueStoryDefault`, but still define the ID as `button--default`, so the prefix wouldn't cause "ugly" IDs.

> Second, it is more convenient to provide the key and have the calling code generate the story ID automatically from the export. And if you provide the ID that can take precedence.

I agree with this, that providing a `key` and auto-generating `name` and `id` from that seems less magical than the other way around.

I propose moving on with this, and if it turns out to be super complicated because a lot of the code assumes `id ~ key` then I'll let you know and backtrack.

Comment (tmeasday) [reply code/lib/types/src/modules/indexer.ts]: I just don't know if we should be adding an extra field to the index. 

If you want to disassociate export name and id, you can just add an id field to the annotations and call the export whatever you like. It's all generated code anyway, right?

Comment (shilman) [reply code/lib/types/src/modules/indexer.ts]: Ok, let me try again. Currently, the indexer must produce an `id`.

Separately, the code that loads the story computes its own `id` here (`key` is called `exportName` here):

https://github.com/storybookjs/storybook/blob/5cb57a87bfc086511b9263855ca7f1bf954cfd21/code/lib/preview-api/src/modules/store/csf/normalizeStory.ts#L55

It's convoluted because the system currently assumes that the ID generation is identical across both codepaths, and if not then it will fail.

The minimal change would be that the Indexer can return a `key`, but that the code that calls the indexer computes the `id` from the `title` and `key` (unless an `id` is also provided), but that the `key` is not stored in the index.

Comment (tmeasday) [reply code/lib/types/src/modules/indexer.ts]: That seems simple enough and uncontroversial. I'm not sure how it interacts with auto title but suspect in a good way.

**Review by tmeasday — APPROVED** [review]
Lgtm

File: code/lib/types/src/modules/indexer.ts (hunk 3 above)
Comment (tobiasdiez) [reply code/lib/types/src/modules/indexer.ts]: This looks like a good design. To make sure I understood it correctly, let's continue the example above: The vue indexer would handle the story `export const vueStoryDefault` by returning `id = default` and `key = vueStoryDefault`. Then the url would use the id, but the loader the key, right?

In that context, it would be nice if the indexer doesn't need to handle id/key conflicts but instead that this would be handled in a central place. For example, the svelte indexer currently keeps track of all ids its encounters and prevents id collisions (e.g. https://github.com/storybookjs/addon-svelte-csf/blob/da2523a272716884667eb6107d6a5d95ab38d706/src/parser/extract-id.ts#L13-L37). In my opinion, this adds unnecessary complications for the indexer (every indexer would need to implement a similar feature) and doesn't work if multiple indexers are used (an indexer doesn't know about the ids that another indexer creates).

Comment (JReinhold) [reply code/lib/types/src/modules/indexer.ts]: > The minimal change would be that the Indexer can return a key, but that the code that calls the indexer computes the id from the title and key (unless an id is also provided), but that the key is not stored in the index.

Now that I've implemented this and thought about it some more, this still doesn't quite make sense to me @shilman.

with this solution you're still computing the export name from the `id` (because that is all there is in the index), so the `id` must directly match the export name. so if your input a `key` that matches the export, but write a custom `id` like `my-pretty-id`, then SB will throw because it can't find the `MyPrettyId` export in the CSF.
I'm not sure what the point is of the indexer being able to specify the `id` if it must match the key anyway.

Comment (JReinhold) [reply code/lib/types/src/modules/indexer.ts]: > In that context, it would be nice if the indexer doesn't need to handle id/key conflicts but instead that this would be handled in a central place. ... In my opinion, this adds unnecessary complications for the indexer (every indexer would need to implement a similar feature) and doesn't work if multiple indexers are used (an indexer doesn't know about the ids that another indexer creates).

Agree @tobiasdiez, this makes sense to me too. Stretch goal @shilman ?
